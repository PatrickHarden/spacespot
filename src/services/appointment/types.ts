export interface AppointmentState {
  data?: Array<Appointment>
  error?: string
}

export interface Appointment {
  id: number
  enquireId: number
  description: string
  peer1Id: string
  peer2Id: string
  proposedTime: string
  createdAt: string
  lastUpdatedAt: string
  lastProposedBy: string
  appointmentStatus: string
}

export interface NewAppointment {
  description: string
  proposedTime: string
}
export interface EditAppointment {
  id: string
  description: string
  proposedTime: string
}
export interface AppointmentAction {
  type: string
  payload?: Array<Appointment> | NewAppointment | EditAppointment | string
}
