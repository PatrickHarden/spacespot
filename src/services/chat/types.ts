export interface ChatAction {
  type: string
  payload?: ChatResponse | ChatMessage | string
}
export interface MessageAction {
  type: string
  payload?: string
}
export interface ChatState {
  data?: ChatResponse
  error?: string
}

export interface ChatResponse {
  chatId: number
  enquiryId: number
  peer1Id: string
  peer2Id: string
  messages: Array<ChatMessage>
}

export interface ChatMessage {
  userId: string
  message: string
  sentAt: string
}
