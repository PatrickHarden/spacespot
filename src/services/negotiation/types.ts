export interface NegotiationTerm {
  termId: number
  value: string
}

export enum TermStatus {
  Pending = 'PENDING_TO_ACCEPT',
  Accepted = 'ACCEPTED',
  Created = 'CREATED',
  Proposed = 'PROPOSED',
  Declined = 'REJECTED',
  Closed = 'CLOSED',
}

export enum DepositType {
  Cash = 'CASH',
  Guarantee = 'GUARANTEE',
}

export enum PaymentType {
  Monthly = 'MONTHLY',
  Quaterly = 'QUARTERLY',
}

export interface NegotiationTerms {
  area?: number
  sizeUnits?: string
  start?: string
  duration?: number
  rent?: number
  serviceCharges?: number
  additionalCosts?: number
  currencyCode?: string
  deposit?: number
  depositType?: DepositType
  frequency?: PaymentType
  noOfDesks?: number
}

// field name capitalized
export interface NegotiationTermFields {
  [key: number]: string
}

// key decapitalized
export interface NegotiationTermFieldsIds {
  [key: string]: number
}

export enum SpecialProvisionStatus {
  Created = 'CREATED',
  Accepted = 'ACCEPTED',
  Proposed = 'PROPOSED',
  Declined = 'REJECTED',
  Closed = 'CLOSED',
  null = 'null',
}

export interface NegotiationAPIResponse {
  terms: Array<NegotiationTerm>
  specialProvision?: string
  specialProvisionStatus?: SpecialProvisionStatus
  status?: TermStatus
  fitoutOption?: number
  peer1SignLeaseStatus?: NegotiationStatus
  peer2SignLeaseStatus?: NegotiationStatus
}

export interface FitoutOptionProps {
  id?: string
  isCustom?: boolean
  name: string
  description: string
  amount: number
}

export interface SelectedFitoutOption {
  custom: boolean
  fitoutOption: any
}

export enum NegotiationStatus {
  TermsNotAgreed = 'TermsNotAgreed',
  TermsAgreed = 'TermsAgreed',
  PreviewStart = 'PreviewStart',
  SignInfo = 'SignInfo',
  AcceptLease = 'AcceptLease',
  LeaseAccepted = 'LeaseAccepted',
  LeaseSignInitiated = 'LeaseSignInitiated',
  LeaseSigned = 'LeaseSigned',
}

export interface NegotiationStatusResponse {
  negotiationId?: number
  peer1SignLeaseStatus: NegotiationStatus
  peer2SignLeaseStatus: NegotiationStatus
}

export interface SignerInfoDTO {
  id?: string
  name: string
  emailId: string
  companyName: string
  companyNumber: string
}

export interface SignerDetailDTO {
  recipientId: string
  name: string
  emailId: string
  signerStatus: 'sent' | 'completed'
  signerType: 'Tenant' | 'Landlord'
  signedDate?: Date
  deliveredDate?: Date
  companyName?: string
  companyNumber?: string
}

export interface NegotiationState {
  data?: NegotiationAPIResponse
  termsFields?: NegotiationTermFields
  customFitout?: Array<FitoutOptionProps>
  signerInfo?: SignerInfoDTO
  error?: string
  loading: boolean
}
