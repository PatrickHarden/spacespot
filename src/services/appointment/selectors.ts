import { get } from 'lodash'
import moment from 'moment'
import { SpacespotState } from '../global/types'
import { Appointment } from './types'
import { AppointmentProps } from '@cbreenterprise/spacespot-ui'

import userSelectors from '../user/selectors'

const appointment = (state: SpacespotState) => get(state, 'appointment.data')
const appointmentsUser = (state: SpacespotState) => {
  const appointments = appointment(state)
  const userId = userSelectors.accountIdentifier(state)
  return appointments
    ? appointments.map(
        (appoint: Appointment): AppointmentProps => ({
          id: appoint.id.toString(),
          date: moment(appoint.proposedTime).toDate(),
          description: appoint.description,
          appointmentStatus: appoint.appointmentStatus,
          isOwnLast: userId === appoint.lastProposedBy,
        }),
      )
    : []
}
const error = (state: SpacespotState) => get(state, 'appointment.error')

export default {
  appointment,
  appointmentsUser,
  error,
}
