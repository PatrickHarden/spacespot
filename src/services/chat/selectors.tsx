import React from 'react'
import styled from 'styled-components'
import { get } from 'lodash'
import moment from 'moment'
import { ChatMsgProps } from '@cbreenterprise/spacespot-ui'
import { SpacespotState } from '../global/types'
import { ChatMessage } from './types'
import eventSelectors from 'services/event/selectors'
import userSelectors from 'services/user/selectors'
import Avatar from 'components/Avatar/Avatar'

const getChat = (state: SpacespotState) => get(state, 'chat.data')
const getChatId = (state: SpacespotState) => get(getChat(state), 'chatId')
const getEnquiryId = (state: SpacespotState) => get(getChat(state), 'enquiryId')
const getPeer1Id = (state: SpacespotState) => get(getChat(state), 'peer1Id')
const getPeer2Id = (state: SpacespotState) => get(getChat(state), 'peer2Id')
const getMessages = (state: SpacespotState): Array<ChatMessage> =>
  get(getChat(state), 'messages')

const ChatAvatar = styled(Avatar)`
  flex: 0 0 auto;
`

const getAvatar = (id: string, isLandlord: boolean) => {
  return <ChatAvatar userId={id} isLandlord={isLandlord} />
}

const getMessagesUser = (state: SpacespotState): Array<ChatMsgProps> => {
  const accountIdentifier = userSelectors.accountIdentifier(state)
  const isLandlord = userSelectors.isLandlord(state)
  const messages = getMessages(state)
  const enquiryId = getEnquiryId(state)
  const events = eventSelectors.getEvents(enquiryId)(state)

  const eventMsgs = events
    ? events.map(e => {
        const isOwn = e.eventMessage.user === accountIdentifier
        const isLandlordMsg = isLandlord ? isOwn : !isOwn
        return {
          ...e,
          id: e.eventDate.toString(),
          isOwn: isOwn,
          time: moment(e.eventDate).toDate(),
          text: '',
          avatar: () => getAvatar(e.eventMessage.user, isLandlordMsg),
        }
      })
    : []

  const userMsgs = messages
    ? messages
        .map(msg => {
          if (!msg.message || msg.message === '') {
            return undefined
          }
          const isOwn = msg.userId === accountIdentifier
          const isLandlordMsg = isLandlord ? isOwn : !isOwn
          return {
            id: msg.sentAt,
            isOwn: isOwn,
            time: moment(msg.sentAt).toDate(),
            text: msg.message,
            avatar: () => getAvatar(msg.userId, isLandlordMsg),
          }
        })
        .filter(msg => msg !== undefined)
    : []

  const sortedMsgs = eventMsgs
    .concat(userMsgs as Array<any>)
    .sort(function(a, b) {
      return a.time.getTime() - b.time.getTime()
    })
    .map((msg: any) => ({
      ...msg,
      time: moment(msg.time).fromNow(),
    }))
  return sortedMsgs
}

const getError = (state: SpacespotState) => get(state, 'chat.error')

export default {
  getChat,
  getChatId,
  getEnquiryId,
  getPeer1Id,
  getPeer2Id,
  getMessages,
  getMessagesUser,
  getError,
}
