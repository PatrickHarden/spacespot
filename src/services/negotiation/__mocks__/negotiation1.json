{
  "terms": [
    { "termId": 1, "value": "34" },
    { "termId": 2, "value": "207" },
    { "termId": 3, "value": "29 Months" },
    { "termId": 4, "value": "200 Sq Mtr" },
    { "termId": 5, "value": "$5600 per month" },
    { "termId": 6, "value": "20th September" }
  ],
  "negotiationId": 129,
  "enquireId": 149,
  "createdAt": "2020-02-03T13:03:07.661+0000",
  "lastUpdated": "2020-02-11T09:06:15.764+0000",
  "lastModifyBy": "c700793a-1963-4b2b-996f-1939f853cc10",
  "status": "CREATED"
}
