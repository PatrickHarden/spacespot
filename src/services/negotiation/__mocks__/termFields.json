{
  "1": "Area",
  "2": "Rent",
  "3": "Start",
  "4": "Duration",
  "5": "Frequency",
  "6": "AdditionalCosts",
  "7": "Deposit",
  "8": "ServiceCharges"
}
