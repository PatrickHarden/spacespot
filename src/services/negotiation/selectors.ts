/* eslint-disable no-unused-vars */
import { get } from 'lodash'
import { SpacespotState } from '../global/types'
import {
  NegotiationTerm,
  NegotiationTerms,
  NegotiationTermFields,
  NegotiationAPIResponse,
  SelectedFitoutOption,
  TermStatus,
  NegotiationTermFieldsIds,
  NegotiationStatus,
  SpecialProvisionStatus,
  SignerInfoDTO,
  SignerDetailDTO,
} from './types'

import { getFields } from './helpers'
import userSelectors from 'services/user/selectors'

const getTerms = (state: SpacespotState): Array<NegotiationTerm> =>
  get(state, 'negotiation.data.terms')

const getStatus = (state: SpacespotState): TermStatus =>
  get(state, 'negotiation.data.status')

const getNegotiationId = (state: SpacespotState): number =>
  get(state, 'negotiation.data.negotiationId')

const getNegotiationFields = (state: SpacespotState): NegotiationTermFields =>
  get(state, 'negotiation.termsFields')

const getNegotiation = (state: SpacespotState): NegotiationAPIResponse =>
  get(state, 'negotiation.data')

const getNegotiationSpecialProvisions = (state: SpacespotState): string =>
  get(getNegotiation(state), 'specialProvision', '')

const decapitalize = (title: string) => title[0].toLowerCase() + title.slice(1)

// FIXME should return undefined if no data loaded to trigger the action
const getNegotiationTerms = (state: SpacespotState): NegotiationTerms => {
  const negTerms = getTerms(state)
  const fields = getNegotiationFields(state)
  const terms = getFields(fields, negTerms)
  return terms
}

const getNegotiationFieldIds = (state: SpacespotState) => {
  const fields = getNegotiationFields(state)
  return (
    fields &&
    Object.keys(fields).reduce((acc, ix) => {
      const i = Number(ix)
      const val = decapitalize(fields[i])
      acc[val] = i
      return acc
    }, {} as NegotiationTermFieldsIds)
  )
}

// FIXME should return undefined if no data loaded to trigger the action
const getNegotiationFitout = (
  state: SpacespotState,
): SelectedFitoutOption | undefined => {
  const fitoutSelected = get(state, 'negotiation.data.fitoutOption')
  if (fitoutSelected === 0) {
    return undefined
  }
  return {
    custom: get(state, 'negotiation.data.customFitout'),
    fitoutOption: fitoutSelected,
  }
}
const getSpecialProvisionStatus = (state: SpacespotState) =>
  get(getNegotiation(state), 'specialProvisionStatus')

const getNegotiationStatus = (state: SpacespotState): NegotiationStatus => {
  const variableName = userSelectors.isLandlord(state)
    ? 'peer2SignLeaseStatus'
    : 'peer1SignLeaseStatus'
  let status = get(getNegotiation(state), variableName)
  if (!status) {
    status =
      getStatus(state) === 'ACCEPTED'
        ? NegotiationStatus.TermsAgreed
        : NegotiationStatus.TermsNotAgreed
  }
  if (status !== NegotiationStatus.TermsAgreed) return status

  const prov = getNegotiationSpecialProvisions(state)
  const provStatus =
    prov && prov !== ''
      ? getSpecialProvisionStatus(state)
      : SpecialProvisionStatus.Accepted

  return provStatus === SpecialProvisionStatus.Accepted
    ? NegotiationStatus.TermsAgreed
    : NegotiationStatus.TermsNotAgreed
}

const getNegotiationPeerStatus = (state: SpacespotState): NegotiationStatus => {
  const variableName = userSelectors.isLandlord(state)
    ? 'peer1SignLeaseStatus'
    : 'peer2SignLeaseStatus'
  const status = get(getNegotiation(state), variableName)
  if (status) {
    return status
  }
  return NegotiationStatus.TermsNotAgreed
}

const getCustomFitout = (
  state: SpacespotState,
): SelectedFitoutOption | undefined => {
  return get(state, 'negotiation.data.customFitout')
}

const getSignerInfo = (state: SpacespotState): SignerInfoDTO | undefined => {
  return get(state, 'negotiation.signerInfo')
}

const getSigners = (
  state: SpacespotState,
): Array<SignerDetailDTO> | undefined => {
  return get(state, 'negotiation.signers')
}

// FIXME in the future we will get the document template Id here
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const getDocumentCulture = (_state: SpacespotState): string => {
  return 'default'
}

export default {
  getNegotiationTerms,
  getTerms,
  getStatus,
  getNegotiation,
  getNegotiationStatus,
  getNegotiationPeerStatus,
  getNegotiationId,
  getNegotiationFields,
  getNegotiationFieldIds,
  getNegotiationFitout,
  getNegotiationSpecialProvisions,
  getCustomFitout,
  getSignerInfo,
  getSigners,
  getDocumentCulture,
}
