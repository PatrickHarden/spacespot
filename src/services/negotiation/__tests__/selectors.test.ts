import { SpacespotState } from '../../global/types'
import selectors from '../selectors'
import {
  NegotiationState,
  NegotiationTermFields,
  NegotiationAPIResponse,
  DepositType,
  PaymentType,
} from '../types'
import negotiation1 from '../__mocks__/negotiation1.json'
import termFields from '../__mocks__/termFields.json'

describe('Negotiation selectors', () => {
  const errorState: SpacespotState = {
    negotiation: { data: {}, loading: false } as NegotiationState,
  }

  const convertJson = (data: any): NegotiationAPIResponse => ({
    ...data,
  })
  const stateNegotiation1: SpacespotState = {
    negotiation: {
      data: convertJson(negotiation1) as NegotiationAPIResponse,
      termsFields: (convertJson(
        termFields,
      ) as unknown) as NegotiationTermFields,
      loading: false,
    } as NegotiationState,
  }

  it('selector should return selected negotiation', () => {
    expect(selectors.getNegotiationTerms(stateNegotiation1)).toMatchSnapshot()
  })
  it('selector should return one without data', () => {
    const defaultTerms = {}
    expect(selectors.getNegotiationTerms(errorState)).toEqual(defaultTerms)
  })
  it('selector should return negotiationId', () => {
    expect(selectors.getNegotiationId(stateNegotiation1)).toEqual(129)
  })
  it('selector should return the terms fields', () => {
    expect(selectors.getNegotiationFields(stateNegotiation1)).toMatchSnapshot()
  })
})
