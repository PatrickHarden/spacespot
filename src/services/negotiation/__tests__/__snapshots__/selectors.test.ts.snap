// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Negotiation selectors selector should return selected negotiation 1`] = `
Object {
  "additionalCosts": "20th September",
  "area": "34",
  "duration": "200 Sq Mtr",
  "frequency": "$5600 per month",
  "rent": "207",
  "start": "29 Months",
}
`;

exports[`Negotiation selectors selector should return the terms fields 1`] = `
Object {
  "1": "Area",
  "2": "Rent",
  "3": "Start",
  "4": "Duration",
  "5": "Frequency",
  "6": "AdditionalCosts",
  "7": "Deposit",
  "8": "ServiceCharges",
}
`;
